From 732db5c89d03fd6c95bfd725247210cbea03739a Mon Sep 17 00:00:00 2001
From: Hilton Chain <hako@ultrarare.space>
Date: Thu, 15 Jan 2026 12:01:21 +0800
Subject: [PATCH] modules: installer: Set up Nonguix.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Don't use ‘check-hardware-support’.
Call %rosenthal-set-keymap script from the installer.
Add Nonguix kernel options.
Set up Nonguix channel in the final system.
Add comment for NVIDIA proprietary driver setup.
WORKAROUND: Set Guile load paths to use modules from external channels.
---
 modules/installer/gnu/installer.scm           | 17 ++++++++-
 modules/installer/gnu/installer/final.scm     |  2 +
 modules/installer/gnu/installer/kernel.scm    | 38 ++++++++++++-------
 .../installer/gnu/installer/newt/kernel.scm   | 10 +++--
 .../installer/gnu/installer/newt/welcome.scm  |  4 +-
 modules/installer/gnu/installer/steps.scm     | 31 ++++++++++++---
 6 files changed, 76 insertions(+), 26 deletions(-)

diff --git a/modules/installer/gnu/installer.scm b/modules/installer/gnu/installer.scm
index adc891e..0f0c766 100644
--- a/modules/installer/gnu/installer.scm
+++ b/modules/installer/gnu/installer.scm
@@ -195,8 +195,21 @@ been performed at build time."
   ;; Apply the specified keymap. Use the default keyboard model.
   #~(match-lambda
       ((layout variant options)
-       (kmscon-update-keymap (default-keyboard-model)
-                             layout variant options))))
+       (begin
+         ;; %rosenthal-set-keymap
+         (system* "/bin/sh" "--login" "-c"
+                  (string-join
+                   `("set-keymap"
+                     ,layout
+                     ,@(or (and=> variant list)
+                           '())
+                     ,@(or (and=> options
+                                  (lambda (x)
+                                    (list "-o" x)))
+                           '()))))
+
+         (kmscon-update-keymap (default-keyboard-model)
+                               layout variant options)))))
 
 (define (compute-keymap-step context dry-run?)
   "Return a gexp that runs the keymap-page of INSTALLER and install the
diff --git a/modules/installer/gnu/installer/final.scm b/modules/installer/gnu/installer/final.scm
index 22c44a2..e88d612 100644
--- a/modules/installer/gnu/installer/final.scm
+++ b/modules/installer/gnu/installer/final.scm
@@ -221,6 +221,8 @@ or #f.  Return #t on success and #f on failure."
              (setvbuf (current-error-port) 'none)
 
              (setenv "PATH" "/run/current-system/profile/bin/")
+             (setenv "GUILE_LOAD_PATH" "/run/current-system/profile/share/guile/site/3.0")
+             (setenv "GUILE_LOAD_COMPILED_PATH" "/run/current-system/profile/lib/guile/3.0/site-ccache")
 
              (for-each (lambda (line)
                          (installer-log-line "~a" line))
diff --git a/modules/installer/gnu/installer/kernel.scm b/modules/installer/gnu/installer/kernel.scm
index 2700931..4e07c32 100644
--- a/modules/installer/gnu/installer/kernel.scm
+++ b/modules/installer/gnu/installer/kernel.scm
@@ -17,6 +17,7 @@
 ;;; along with GNU Guix.  If not, see <http://www.gnu.org/licenses/>.
 
 (define-module (gnu installer kernel)
+  #:use-module (ice-9 match)
   #:use-module (gnu system hurd)
   #:use-module (guix read-print)
   #:export (kernel->configuration))
@@ -27,16 +28,27 @@
   str)
 
 (define (kernel->configuration kernel dry-run?)
-  (if (and (string? kernel)
-           (string-prefix? "Hurd" kernel))
-      `((kernel %hurd-default-operating-system-kernel)
-        ,(comment (G_ ";; \"noide\" disables the gnumach IDE driver, enabling rumpdisk.\n"))
-        (kernel-arguments '("noide"))
-        (firmware '())
-        (hurd hurd)
-        (locale-libcs (list glibc/hurd))
-        (name-service-switch #f)
-        (essential-services (hurd-default-essential-services this-operating-system))
-        (privileged-programs '())
-        (setuid-programs %setuid-programs/hurd))
-      '()))
+  (match kernel
+    ("Linux"
+     `((kernel linux)
+       (firmware (cons* linux-firmware %base-firmware))
+       (initrd microcode-initrd)))
+    ("Linux long-term support (LTS)"
+     `((kernel linux-lts)
+       (firmware (cons* linux-firmware %base-firmware))
+       (initrd microcode-initrd)))
+    ((? (lambda (str)
+          (and (string? str)
+               (string-prefix? "Hurd" str))))
+     `((kernel %hurd-default-operating-system-kernel)
+       ,(comment (G_ ";; \"noide\" disables the gnumach IDE driver, enabling rumpdisk.\n"))
+       (kernel-arguments '("noide"))
+       (firmware '())
+       (hurd hurd)
+       (locale-libcs (list glibc/hurd))
+       (name-service-switch #f)
+       (essential-services (hurd-default-essential-services this-operating-system))
+       (privileged-programs '())
+       (setuid-programs %setuid-programs/hurd)))
+    (_
+     '())))
diff --git a/modules/installer/gnu/installer/newt/kernel.scm b/modules/installer/gnu/installer/newt/kernel.scm
index 4570544..6e64bc4 100644
--- a/modules/installer/gnu/installer/newt/kernel.scm
+++ b/modules/installer/gnu/installer/newt/kernel.scm
@@ -26,9 +26,13 @@
   ;; TRANSLATORS: "Hurd" is a proper noun and must not be translated.
   (let* ((hurd-x86 (G_ "Hurd 32-bit (experimental)"))
          (hurd-x86_64 (G_ "Hurd 64-bit (highly experimental!)"))
+         (linux "Linux")
+         (linux-lts "Linux long-term support (LTS)")
          (linux-libre "Linux Libre")
          (kernels (parameterize ((%current-target-system #f))
-                    `(,linux-libre
+                    `(,linux
+                      ,linux-lts
+                      ,linux-libre
                       ,@(cond ((target-x86-64?)
                                (list hurd-x86 hurd-x86_64))
                               ((target-x86?)
@@ -40,14 +44,14 @@
                         ((equal? (%current-target-system) "x86_64-pc-gnu")
                          hurd-x86_64)
                         (else
-                         linux-libre)))
+                         linux)))
          (result
           (run-listbox-selection-page
            #:title (G_ "Kernel")
            #:info-text
            ;; TRANSLATORS: "Hurd" is a proper noun and must not be translated.
            ;; TRANSLATORS: "Linux Libre" is a literal and must not be translated.
-           (G_ "Please select a kernel.  When in doubt, choose \"Linux Libre\".
+           (G_ "Please select a kernel.  When in doubt, choose \"Linux\".
 
 The Hurd is offered as a technology preview and development aid; many packages \
 are not yet available in Guix, such as a desktop environment or even a \
diff --git a/modules/installer/gnu/installer/newt/welcome.scm b/modules/installer/gnu/installer/newt/welcome.scm
index f821374..fddcc93 100644
--- a/modules/installer/gnu/installer/newt/welcome.scm
+++ b/modules/installer/gnu/installer/newt/welcome.scm
@@ -184,12 +184,10 @@ Documentation is accessible at any time by pressing Ctrl-Alt-F2.")
    #:listbox-items
    `((,(G_ "Graphical install using a terminal based interface")
       .
-      ,(lambda ()
-         (check-hardware-support pci-database)))
+      ,(const #t))
      (,(G_ "Install using the shell based process")
       .
       ,(lambda ()
-         (check-hardware-support pci-database)
          ;; Switch to TTY3, where a root shell is available for shell based
          ;; install. The other root TTY's would have been ok too.
          (system* "chvt" "3")
diff --git a/modules/installer/gnu/installer/steps.scm b/modules/installer/gnu/installer/steps.scm
index b70e5eb..70e8010 100644
--- a/modules/installer/gnu/installer/steps.scm
+++ b/modules/installer/gnu/installer/steps.scm
@@ -242,13 +242,34 @@ found in RESULTS."
 ;; Indicate which modules to import to access the variables
 ;; used in this configuration.\n"))
                    ,@(if (target-hurd?)
-                         '((use-modules (gnu) (gnu system hurd))
-                           (use-package-modules hurd ssh))
-                         '((use-modules (gnu))))
-                   (use-service-modules cups desktop networking ssh xorg))))
+                         '((use-modules (gnu)
+                                        (nonguix)
+                                        (gnu system hurd)
+                                        (gnu services cups)
+                                        (gnu services desktop)
+                                        (gnu services networking)
+                                        (gnu services ssh)
+                                        (gnu services xorg)
+                                        (gnu packages hurd)
+                                        (gnu packages ssh)))
+                         '((use-modules (gnu)
+                                        (nonguix)
+                                        (gnu services cups)
+                                        (gnu services desktop)
+                                        (gnu services networking)
+                                        (gnu services ssh)
+                                        (gnu services xorg)))))))
     `(,@modules
       ,(vertical-space 1)
-      (operating-system ,@configuration))))
+      (define %my-os
+        (operating-system ,@configuration))
+      ,(vertical-space 1)
+      ((compose
+        ,(comment "\
+;; Uncomment the following line for NVIDIA proprietary driver support.
+;; (nonguix-transformation-nvidia #:configure-xorg? #t)\n")
+        (nonguix-transformation-guix))
+       %my-os))))
 
 (define* (configuration->file configuration
                               #:key (file-name (%installer-configuration-file)))
-- 
2.52.0

